File: view/js/list.js

wab.List.prototype.delete = function(ref) {
    wab.httpCall('DELETE', '/v1/' + this.kind + '/' + ref, this, function(ol, resp) {
        wab.view.set(ol);
    })
}

wab.List.prototype.display = function(view, edit) {
    var wrapper = wab.classifyNewElement('div', 'table-wrapper');
    view.appendChild(wrapper);

    e = wab.classifyNewElement('div', 'btn');

    btn = document.createElement('span');
    btn.appendChild(document.createTextNode('Create'));

    (function(ol) { e.onclick = function() { wab.view.set(new wab.Obj(0, ol), true); }})(this);
    e.appendChild(btn);
    wrapper.appendChild(e);

    var frame = document.createElement('table'), table, row, cell;
    wrapper.appendChild(frame);

    row = document.createElement('tr');
    frame.appendChild(row);

    header = wab.classifyNewElement('table', 'obj-list-table');
    row.appendChild(header);

    row = document.createElement('tr');
    header.appendChild(row);

    var cs, len = this.spec.list.length, opt = '/list?'
    for (i = 0; i < len; i++) {
        cs = this.spec.list[i];
        if (0 < i) {
            opt += '&'
        }
        opt += cs.key + '=' + cs.path;
        cell = document.createElement('th');
        cell.appendChild(document.createTextNode(cs.label));
        row.appendChild(cell);
    }
    cell = wab.classifyNewElement('th', 'list-actions');
    cell.appendChild(document.createTextNode('Actions'));
    cell.setAttribute('colspan', 3);
    row.appendChild(cell);

    // Prepare list table.
    row = wab.classifyNewElement('tr', 'list-items');
    frame.appendChild(row);
    table = wab.classifyNewElement('table', 'obj-list-table');
    row.appendChild(table);

    this.table = table;
    
    // Request content.
    wab.httpCall('GET', '/v1/' + this.kind + opt, this, function(list, resp) {
        var results = resp.body.results, btn, bi;
        if (typeof results === 'object') {
            var i, cs, len = list.spec.list.length;
            var j, obj, rlen = results.length, ref;
            for (j = 0; j < rlen; j++) {
                obj = results[j];
                ref = obj.ref;
                row = document.createElement('tr');
                list.table.appendChild(row);
                for (i = 0; i < len; i++) {
                    cs = list.spec.list[i];
                    cell = wab.classifyNewElement('td', 'obj-list');
                    cell.appendChild(document.createTextNode(obj[cs.key]));
                    row.appendChild(cell);
                }
                for (i = 0; i < wab.listButtons.length; i++) {
                    bi = wab.listButtons[i];
                    cell = wab.classifyNewElement('td', bi.cn);
                    btn = wab.classifyNewElement('span', bi.icon);
                    btn.setAttribute('title', bi.title);
                    cell.appendChild(btn);
                    // A function is needed to copy the variables.
                    (function(list, r, b, act) {
                        switch(act) {
                        case 2:
                            b.onclick = function() { list.delete(r); }
                            break;
                        case 1:
                            b.onclick = function() { wab.view.set(new wab.Obj(r, list), true); }
                            break;
                        case 0:
                        default:
                            b.onclick = function() { wab.view.set(new wab.Obj(r, list), false); }
                        }
                    })(list, ref, btn, bi.act);
                    row.appendChild(cell);
                }
            }
        }
    })
}

File: lib/wab/ui/list.rb

module WAB
  module UI

    # Represents a list display.
    class List < Display

      attr_accessor :list_paths
      
      def initialize(kind, id, template, list_paths, transitions)
        super(kind, id, template, transitions, 'ui.List')
        @list_paths = list_paths
      end

      def spec
        ui_spec = super
        ui_spec[:table] = html_table
        ui_spec[:row] = html_row
        ui_spec
      end

      # Returns an HTML string to be used as the table of a list of
      # objects. The table must have an +id+ attribute value of the +name+
      # argument. Generally the column header should include the list_paths or
      # more friendly alternatives. If a create button is desired then an
      # element with the name joined with '.create_button' should be the +id+ of
      # the element.
      def html_table
        html = %{<div class="table-wrapper"><div>#{@kind} List</div><div class="btn" style="float: right;" id="#{@name}.create_button"><span>Create</span></div><table class="obj-list-table" id="#{@name}.table"><tr>}
        # The column headers.
        @list_paths.map { |path| html << "<th>#{path.capitalize}</th>" }
        # Add the view, edit, and delete buttons header.
        html << %{<th colspan="3">Actions</th>}
        html << %{</tr></table></div>}
      end

      # Returns an HTML string to be used as a row in a table of a list of
      # objects. Each column element should include the list_paths identifier
      # as a +path+ element attribute.
      #
      # Each row can contain the buttons +view_button+, +edit_button+, and
      # +delete_button+. These will be used to set up the transitions.
      def html_row
        html = '<tr>'
        @list_paths.map { |path| html << %{<td class="obj-list" path="#{path}"></td>} }
        buttons = [
                   { title: 'View', icon: 'icon icon-eye', cn: 'actions' },
                   { title: 'Edit', icon: 'icon icon-pencil', cn: 'actions' },
                   { title: 'Delete', icon: 'icon icon-trash-o', cn: 'actions delete' }
                  ].map { |spec|
          html << %{<td class="#{spec[:cn]}"><span class="#{spec[:icon]}" title="#{spec[:title]}"></span></td>}
        }
        html << '</tr>'
      end

    end # List
  end # UI
end # WAB

